File: packages/entity-sync-core/src/codec.ts
/**
 * MessagePack encode/decode for PSyncV1 messages — the TypeScript
 * counterpart to `pes-protocol/src/codec.rs`.
 *
 * Uses `@msgpack/msgpack`, which encodes plain JS objects as MessagePack
 * maps (matching Rust's `rmp_serde::to_vec_named`) and Rust's externally-
 * tagged enum representation, per `messages.ts`'s doc comment. No custom
 * encoder/decoder config is needed: the wire shapes line up because both
 * sides use their language's natural default representation for
 * map-of-string-keys and string/bytes types.
 */

import { encode, decode } from "@msgpack/msgpack";
import type { ClientMessage, ServerMessage } from "./messages.js";

/** Encode a {@link ServerMessage} to MessagePack bytes. */
export function encodeServer(msg: ServerMessage): Uint8Array {
  return encode(msg);
}

/** Encode a {@link ClientMessage} to MessagePack bytes. */
export function encodeClient(msg: ClientMessage): Uint8Array {
  return encode(msg);
}

/**
 * Decode a {@link ServerMessage} from MessagePack bytes.
 *
 * Forward-compatible: unrecognized map keys in `bytes` (e.g. fields added
 * by a future protocol version) are silently ignored by
 * `@msgpack/msgpack`'s default decode behavior, matching the Rust side's
 * `serde` default (no `deny_unknown_fields`).
 */
export function decodeServer(bytes: Uint8Array): ServerMessage {
  return decode(bytes) as ServerMessage;
}

/**
 * Decode a {@link ClientMessage} from MessagePack bytes. See
 * {@link decodeServer} for the forward-compatibility contract.
 */
export function decodeClient(bytes: Uint8Array): ClientMessage {
  return decode(bytes) as ClientMessage;
}

File: crates/pes-protocol/src/codec.rs
//! MessagePack encode/decode for PSyncV1 messages.
//!
//! Uses `rmp_serde::to_vec_named` (not `to_vec`) so every message is
//! encoded as a MessagePack map keyed by field name, not a positional
//! array. This is what makes forward compatibility possible: a decoder
//! reading an older or newer message shape simply ignores map keys it
//! doesn't recognize (serde's default behavior — no `deny_unknown_fields`
//! is ever set on these types), rather than misreading positional fields
//! that shifted when a field was added or removed.

use bytes::Bytes;

use crate::error::ProtocolError;
use crate::messages::{ClientMessage, ServerMessage};

/// Encode a [`ServerMessage`] to MessagePack bytes.
pub fn encode_server(msg: &ServerMessage) -> Result<Bytes, ProtocolError> {
    let bytes = rmp_serde::to_vec_named(msg)?;
    Ok(Bytes::from(bytes))
}

/// Encode a [`ClientMessage`] to MessagePack bytes.
pub fn encode_client(msg: &ClientMessage) -> Result<Bytes, ProtocolError> {
    let bytes = rmp_serde::to_vec_named(msg)?;
    Ok(Bytes::from(bytes))
}

/// Decode a [`ServerMessage`] from MessagePack bytes.
///
/// Forward-compatible: unrecognized map keys in `bytes` (e.g. fields added
/// by a future protocol version) are silently ignored rather than causing
/// a decode error.
pub fn decode_server(bytes: &[u8]) -> Result<ServerMessage, ProtocolError> {
    Ok(rmp_serde::from_slice(bytes)?)
}

/// Decode a [`ClientMessage`] from MessagePack bytes.
///
/// Forward-compatible: unrecognized map keys in `bytes` are silently
/// ignored rather than causing a decode error.
pub fn decode_client(bytes: &[u8]) -> Result<ClientMessage, ProtocolError> {
    Ok(rmp_serde::from_slice(bytes)?)
}
